### PR TITLE
Add raw mmap

### PR DESCRIPTION
**Pull request by diwic**
Hi!

In order to use the memmap2 crate for my [IPC experiments](https://github.com/diwic/shmem-ipc), I had to create a raw mmap which just exposes a pointer and a length, no slices.

Due to Rust's aliasing rules, creating slices to memory can be unsound if more than one process can write to the same memory area. (The IPC protocol can dictate which parts of the memory that can be accessed in what way.)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
...
+/// This struct never hands out references to its interior, only raw pointers.
+/// This can be helpful when creating shared memory maps between untrusted processes.
+pub struct MmapRaw {
+    inner: MmapInner,
+}
+
+impl MmapRaw {
+    /// A raw pointer to the interior.
+    ///
+    /// ## Safety
```
Comment (RazrFalcon) [inline src/lib.rs:493]: One `#` is enough.

File: src/lib.rs [hunk 2]
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
...
+    ///
+    /// To safely dereference this pointer, you need to make sure that the file has not been
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
+    /// The length in bytes of the memory map.
```
Comment (RazrFalcon) [inline src/lib.rs:499]: New line.

File: src/lib.rs [hunk 3]
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
...
+
+impl MmapRaw {
+    /// A raw pointer to the interior.
+    ///
+    /// ## Safety
+    ///
+    /// To safely dereference this pointer, you need to make sure that the file has not been
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
```
Comment (RazrFalcon) [inline src/lib.rs:498]: This should return `*const u8`. And `*mut u8` should be under `as_mut_ptr`.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by diwic** [comment]
Thanks for the quick review! I've pushed another commit that addresses your feedback. Let me know if you have any further thoughts.

After merging, I would appreciate a crates.io release (so I can publish my crate on crates.io). Thanks in advance!

File: src/lib.rs [hunk 4]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+    /// # Errors
+    ///
+    /// This method returns an error when the underlying system call fails, which can happen for a
+    /// variety of reasons, such as when the file is not open with read and write permissions.
+    /// ```
+    pub fn map_raw(file: &File) -> Result<MmapRaw> {
+        MmapOptions::new().map_raw(file)
+    }
+
+    /// A raw mutable pointer to the interior.
```
Comment (RazrFalcon) [inline src/lib.rs:504]: Rename to _Returns a raw pointer to the memory mapped file._ To match [slice::as_ptr](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.as_ptr)

File: src/lib.rs [hunk 5]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+    ///
+    /// # Safety
+    ///
+    /// To safely dereference this pointer, you need to make sure that the file has not been
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_ptr(&self) -> *const u8 { self.inner.ptr() }
+
+    /// A raw mutable pointer to the interior.
```
Comment (RazrFalcon) [inline src/lib.rs:513]: Rename to _Returns an unsafe mutable pointer to the memory mapped file._

File: src/lib.rs [hunk 6]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+/// This struct never hands out references to its interior, only raw pointers.
+/// This can be helpful when creating shared memory maps between untrusted processes.
+pub struct MmapRaw {
+    inner: MmapInner,
+}
+
+impl MmapRaw {
+    /// Creates a writeable memory map backed by a file.
+    ///
+    /// This is equivalent to calling `MmapOptions::new().map_mut(file)`.
```
Comment (RazrFalcon) [inline src/lib.rs:493]: Change to `MmapOptions::new().map_raw(file)`

File: src/lib.rs [hunk 7]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+    /// A raw mutable pointer to the interior.
+    ///
+    /// # Safety
+    ///
+    /// To safely dereference this pointer, you need to make sure that the file has not been
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_mut_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
+
+    /// The length in bytes of the memory map.
```
Comment (RazrFalcon) [inline src/lib.rs:522]: Rename to _Returns the length in bytes of the memory map._

File: src/lib.rs [hunk 8]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+    ///
+    /// Note that truncating the file can cause the length to change (and render this value unusable).
+    #[inline]
+    pub fn len(&self) -> usize { self.inner.len() }
+}
+
+impl fmt::Debug for MmapRaw {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Mmap")
```
Comment (RazrFalcon) [inline src/lib.rs:531]: `Mmap` -> `MmapRaw`

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by diwic** [comment]
Alright, second round fixed now. Thanks for the quick review!

File: src/lib.rs [hunk 9]
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
...
+impl MmapRaw {
+    /// Creates a writeable memory map backed by a file.
+    ///
+    /// This is equivalent to calling `MmapOptions::new().map_raw(file)`.
+    ///
+    /// # Errors
+    ///
+    /// This method returns an error when the underlying system call fails, which can happen for a
+    /// variety of reasons, such as when the file is not open with read and write permissions.
+    /// ```
```
Comment (RazrFalcon) [inline src/lib.rs:499]: Remove this line.

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by RazrFalcon** [comment]
Almost there.

**Comment by diwic** [comment]
Fixed and squashed into one commit. Sorry about the sloppiness - comments are the hardest to get right, because the compiler does not check them for you ;-)

**Comment by RazrFalcon** [comment]
No problem.

**Comment by RazrFalcon** [comment]
Published.

**Comment by diwic** [comment]
Thanks!

**Pull request merged**